### ISSUE TITLE
SciPy for Biomedicine

### ISSUE THREAD

**Issue by mdhaber** [scipy/scipy#16191]
This is a meta-issue recording action items from a review of biomedical articles that cite SciPy and the results of a survey sent to authors of over 50 of these papers.

- [ ] [Vaccine optimization for COVID-19: Who to vaccinate first?](https://www.science.org/doi/full/10.1126/sciadv.abf1374) used `optimize.minimize` to study how COVID vaccination priority should be assigned. The researchers reported having to “repair” solutions returned by SciPy’s Nelder-Mead solver, but had the researchers used a solver capable of enforcing problem constraints, this might have been avoided. To resolve this, we will improve the documentation of the optimizers to ensure that researchers find the appropriate solver for their problem, and we will enhance the solvers’ support for constraints.
- [ ] [Modeling and forecasting the early evolution of the Covid-19 pandemic in Brazil](https://www.nature.com/articles/s41598-020-76257-1) used the `solve_ivp` method to numerically integrate a system of ordinary differential equations. Leveraging Boost, the C++ library added to SciPy as part of EOSS-0000000432, we will add more modern integrators to `solve_ivp` to improve `solve_ivp`’s performance.
- [ ] [Multi-Omics Resolves a Sharp Disease-State Shift between Mild and Moderate COVID-19](https://www.sciencedirect.com/science/article/pii/S0092867420314446) used the `hierarchy.fcluster` function to perform T cell receptor (TCR) clustering analysis. We will resolve the [two open bug reports](https://github.com/scipy/scipy/issues?q=is%3Aissue+is%3Aopen+fcluster+) about `fcluster` to facilitate future use.
- [ ] An author of [The impact of reducing signal acquisition specifications on neuronal spike sorting](https://ieeexplore.ieee.org/document/9630669) suggested "having helper functions within scipy.signal.fft that 1) converts the fft coefficients to power (V**2/Hz) and 2) converts power and phase to fft coefficients would be nice. I always forget how to do the conversion and a function that does this would be nice to have". 
- [ ] The same author suggested improving the speed of [scipy.signal.resample](https://docs.scipy.org/doc/scipy/reference/generated/scipy.signal.resample.html) and "having an option to resample via a non-Fourier method in the case the signal is not periodic"
- [ ] An author of [COVID-19 pandemic-related lockdown: response time is more important than its strictness](https://www.embopress.org/doi/full/10.15252/emmm.202013171) reported about [scipy.stats.curve_fit](https://docs.scipy.org/doc/scipy/reference/generated/scipy.optimize.curve_fit.html): "Non-linear optimizations is not trivial, as you have to tweak parameters (tolerances, jacobians parameters, methods of optimization). If you could automate this parameters choice, it would be helpful." An author of [Inferring gene regulatory networks from single-cell RNA-seq temporal snapshot data requires higher- order moments](https://www.sciencedirect.com/science/article/pii/S266638992100180X) had similar comments, suggesting the use of stochastic gradient descent/global optimization. An author of [A General Model to Calculate the Spin–Lattice Relaxation Rate (R1) of Blood](https://onlinelibrary.wiley.com/doi/full/10.1002/jmri.27938) also reported using `curve_fit`, noting that "It was super easy to use".
- [ ] An author of [pymia: A Python package for data handling and evaluation in deep learning-based medical image analysis](https://www.sciencedirect.com/science/article/pii/S0169260720316291) suggested "I would love to have some of the ndimage functions working on pytorch or tensorflow tensors. Often, I need to cast to np.ndarry just to use the SciPy functions."
- [ ] An author of [Quantitative imaging of apoptosis following oncolytic virotherapy by magnetic resonance fingerprinting aided by deep learning](https://www.nature.com/articles/s41551-021-00809-7) mentioned "Generally, I think that enriching the documentation with additional practical examples and perhaps links to relevant examples or papers that used the functions could be helpful."
- [ ] An author of [Covasim: An agent-based model of COVID-19 dynamics and interventions](https://journals.plos.org/ploscompbiol/article?id=10.1371/journal.pcbi.1009149) said "We noticed that scipy.stats is a bit slow to import -- about 0.3 s. For this reason we import it only within the function rather than globally, to reduce Covasim load time. If there were a way to reduce the import time, that would be appreciated." We will address this by overhauling the univariate distribution infrastructure so that distributions are not instantiated at load time.
- [ ] An author of [Empirical mode decomposition for automatic artifact elimination in electrogastrogram](https://ieeexplore.ieee.org/document/9400683) noted "it would be great to add asymmetrical windows. I don't mean about sym parameter in signal.windows.blackman as it generates either symmetric window or periodic window, but to the design of a window."
- [ ] The same author noted "For example there are excellent methods in SciPy as resample, which I need very often, but I'm not aware of them and have never used them. If I may add, it might be helpful to have more tutorials online, not just for research, but for education as well."
- [ ] An author of [TorchIO: A Python library for efficient loading, preprocessing, augmentation and patch-based sampling of medical images in deep learning](https://arxiv.org/abs/2003.04696) suggested "From my point of view, the obvious enhancement that would facilitate biomedical research is adding support to perform operations in physical space, i.e., assuming that images have a certain origin, orientation and spacing."
- [ ] An author of [Modeling COVID-19 Dynamics in Illinois under Nonpharmaceutical Interventions](https://journals.aps.org/prx/abstract/10.1103/PhysRevX.10.041033) wrote that "...stochastic differential equation solvers might be a valuable addition..." but noted that "it might require a level of effort and domain expertise beyond the justifiable scope of SciPy itself." Stochastic differential equation solvers would indeed be useful and probably in scope, so we will look into the feasibility of including one (likely vendored) in SciPy.

The following are action items from a review of biomedical software that relies on SciPy and the results of a survey sent to maintainers of these packages.
- [ ] Napari’s Octree feature uses `scipy.ndimage.zoom`, but the Napari source code comments ([here ](https://github.com/napari/napari/blob/de4a2f21f30210e660a5d9c2142da3b0133e28a2/napari/layers/image/image.py#L834-L837)and [here](https://github.com/napari/napari/blob/5f96d5d814aad697c367bdadbb1a57750e2114ad/napari/layers/image/experimental/octree_tile_builder.py#L78-L80)) that `zoom` does not support the `float16` data type. We will add support for `float16` and other memory-efficient data types where appropriate throughout `scipy.ndimage`.
- [ ] There are two issues on Napari’s issue tracker (napari/napari#927, napari/napari#1347) that note the need to `import scipy.stats` to fix a mysterious bug, but the cause of the problem has not been identified. We will determine what in `scipy.stats` fixes the bug to ensure that changes to `scipy.stats` do not break Napari.
- [ ] Napari [uses](https://github.com/napari/napari/blob/e11085c3b197d36ad35612966bbf94b810026b76/examples/affine_transforms.py#L33-L38) `scipy.ndimage.affine_transform`, but the Napari source code notes that SciPy’s “pull” style of affine transforms requires the transformation matrix to be inverted before it can be used by Napari. We will add native support for Napari’s “push” transforms to make use of `affine_transform` more convenient and efficient.
- [ ] It was noted that the API could be improved by having more explicit parameter names. ImageJ was said to be more intuitive, so we will review its API. API stability being a real concern, we will make changes with care and respecting our deprecation policy.
- [ ] More practical examples relevant to the biomedical community are wanted. In general, our documentation is seen as complicated to read with lot of technical language. We will create introduction content explaining main concepts. E.g. in scipy.stats we will explain how to use which tests and introduce the testing vocabulary (p-value, power, statistics, etc.).
- [ ] There are still blockers for people to contribute and SciPy is still seen as a difficult project to contribute to. We will work on improving our communication and on-boarding of newcomers. One of the area we want to explore is to have pair programming between domain experts and SciPy maintainers. It would help non professional programmers to contribute.
- [ ] Scikit-image is fairly used for filtering and segmentation. The features rely on SciPy but are mostly limited to 2D. We will review our API to make direct use of our functions easier (and provide more documentation), and also investigate potential blockers on SciPy's side that prevent downstream libraries like Scikit-image to use the full potential of SciPy.

**Comment by mdhaber** [comment]
Even when I follow links to open issues against my fork of SciPy, they get opened against the main repo. Oh well.

**Comment by drammock** [comment]
+1, thanks for prioritizing downstream wishes and pain points!  These changes would be great to see.

**Comment by orbeckst** [comment]
Thank you scipy team for maintaining the scipy package, a bed-rock of scientific computing. Without a doubt, scipy is **essential open source scientific software** and we the [@MDAnalysis/coredevs](https://github.com/orgs/MDAnalysis/teams/coredevs) are happy to support the scipy project.

In [MDAnalysis](https://www.mdanalysis.org), a Python package for the analysis of particle-based simulations such as Molecular Dynamics simulations of biomolecules. MDAnalysis was recently funded under CZI-EOSS4.

We rely on a number of scipy functions for our built-in analysis tools (e.g., `scipy.optimize` for curve fitting, `scipy.stats.gaussian_kde` for density representations, `scipy.integrate.simps` for integration, `scipy.spatial.distance.directed_hausdorff()` for path similarity analysis, ...) and users of our packages use scipy downstream from us (e.g., for clustering, spectral analysis, ...). We also use `scipy.sparse` to represent some sparse distance matrices.

In the core library, we use `scipy.spatial.cKDTree` for neighbor searching and distance matrices within a cutoff. We augment `cKDTree` (see [PeriodicKDTree](https://docs.mdanalysis.org/stable/documentation_pages/lib/pkdtree.html#MDAnalysis.lib.pkdtree.PeriodicKDTree)) to be able to deal with 3D-periodic systems, which are common in molecular simulations but for larger systems the performance is not as good as our own [capped_distance()](https://docs.mdanalysis.org/stable/documentation_pages/lib/distances.html#MDAnalysis.lib.distances.capped_distance) grid-based search. In general, native support for periodic systems (in orthorhombic and triclinic unit cells) would be really useful for us (and might also be useful for biomolecular X-ray crystallography applications).

We also use `scipy.io.netcdf` as a light-weight NCDF reader. The reading performance is very good but writing is so slow that we prefer to fall back to the netcdf library (see also #9157 and issues linked there); improving the writing performance would allow us to remove a heavy dependency and always provide good performance to users.

We'd be happy to provide you with more information if needed.

**Comment by jeylau** [comment]
At @DeepLabCut we make extensive use of the optimize and spatial modules of SciPy.
The first is mostly for solving optimal assignment problems with linear_sum_assignment, whereas the latter is for neighbor search (2D/3D mainly), pairwise distance computation, etc., with cKDTree, pdist, cdist, squareform, euclidean, directed_hausdorff; most of our code dealing with point coordinates relies on these SciPy functions really!

Note that we do use other modules too, but to a lesser extent:
signal.medfit and butter and interpolate.CubicSpline for time series smoothing/filtering;
linalg.interpolative and linalg.hankel for matrix low-rank approximation and time series delay embedding respectively;
stats.gaussian_kde and chi2 to determine prior distributions;
io.savemat for backwards compatibility with data from early projects stored in .mat;
ndimage.measurements for local peak detection.

They all work very smoothly, though 2 things come to mind: we found ourselves often wanting to fit splines over the columns of a 2D array (e.g., multivariate time series), but without an « axis » argument, we resort to np.stack and a for loop; the same minor thing could apply to ndimage routines, so batches of images can be readily passed in.

Overall though, SciPy is fantastic, and we'd see the support of large tensors already on GPU and routines with multiple workers becoming crucial to biomedical research.

**Comment by tupui** [comment]
Thank you everyone for your valuable inputs 😃 For anyone reading we are keeping the survey open: https://forms.gle/Du9Y1aZrDx37LY8Y9
Feel free to comment here as well.

**Comment by mdhaber** [comment]
@arliss-NF

**Comment by ev-br** [comment]
> we use scipy.spatial.cKDTree for neighbor searching and distance matrices within a cutoff. We augment cKDTree (see [PeriodicKDTree](https://docs.mdanalysis.org/stable/documentation_pages/lib/pkdtree.html#MDAnalysis.lib.pkdtree.PeriodicKDTree)) to be able to deal with 3D-periodic systems,

@orbeckst is there a deep reason MDAnalysis does not use the cKDTree's `boxsize` parameter? https://docs.scipy.org/doc/scipy/reference/generated/scipy.spatial.cKDTree.html

**Comment by orbeckst** [comment]
I think `boxsize` os recent compared to our code, so we were probably not aware of it. We should be able to use it for orthorhombic boxes. 

For triclinic boxes there is no alternative to augmentation (explicitly adding the images), at least to my knowledge.

**Comment by dschmitz89** [comment]
Can we close this since all boxes are ticked off?

**Comment by mdhaber** [comment]
I think so, if all the resolutions look reasonable to you? We weren't able to satisfy all requests, but I think we made some progress and gave fair responses to the rest. If there are some things that deserve further investigation, we can open separate issues.
